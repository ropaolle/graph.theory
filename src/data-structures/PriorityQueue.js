class Node {
  constructor(val, priority) {
    this.val = val;
    this.priority = priority;
  }
}

export default class PriorityQueue {
  constructor() {
    this.items = [];
  }

  get isEmpty() {
    return this.items.length === 0;
  }

  get size() {
    return this.items.length;
  }

  enqueue(val, priority) {
    var element = new Node(val, priority);
    var contain = false;
    // Iterating through items and add elements at the correct location of the Queue.
    for (var i = 0; i < this.items.length; i++) {
      if (this.items[i].priority > element.priority) {
        // Once the correct location is found it is enqueued.
        this.items.splice(i, 0, element);
        contain = true;
        break;
      }
    }

    // Highest priority element is added to queue.
    if (!contain) this.items.push(element);
  }

  dequeue() {
    return this.items.shift();
  }

  front() {
    return this.items[0];
  }

  rear() {
    return this.items[this.items.length - 1];
  }

  toString(includePriority = false) {
    let items = [];
    for (const { val, priority } of this.items)
      items.push(includePriority ? `${val}:${priority}` : val);

    return items.join(' -> ');
  }

  /* istanbul ignore next */ print() {
    console.info(this.toString());
  }
}
