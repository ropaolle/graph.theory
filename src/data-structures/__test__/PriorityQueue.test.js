import PriorityQueue from '../PriorityQueue.js';

describe('BinaryTree (number)', () => {
  const pq = new PriorityQueue();

  it('should enqueue items correctly', () => {
    pq.enqueue('A', 1);
    pq.enqueue('B', 2);
    pq.enqueue('C', 3);
    pq.enqueue('1', 10);
    pq.enqueue('2', 10);
    pq.enqueue('3', 10);
    pq.enqueue('X', 5);
    pq.enqueue('Y', 5);
    pq.enqueue('Z', 5);
  });

  it('should access front and rear item correctly', () => {
    expect(pq.front().val).toBe('A');
    expect(pq.rear().val).toBe('3');
  });

  it('render toString correctly', () => {
    expect(pq.toString()).toMatchSnapshot();
    expect(pq.toString(true)).toMatchSnapshot();
  });

  it('should dequeue items correctly', () => {
    expect(pq.dequeue()).toMatchObject({ priority: 1, val: 'A' });
    expect(pq.dequeue()).toMatchObject({ priority: 2, val: 'B' });
    expect(pq.isEmpty).toBe(false);
    expect(pq.size).toBe(7);
    pq.dequeue();
    pq.dequeue();
    pq.dequeue();
    pq.dequeue();
    pq.dequeue();
    pq.dequeue();
    pq.dequeue();
    expect(pq.isEmpty).toBe(true);
    expect(pq.dequeue()).toBeUndefined();
  });
});
