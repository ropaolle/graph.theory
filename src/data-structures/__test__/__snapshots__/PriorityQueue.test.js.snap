// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinaryTree (number) render toString correctly 1`] = `"A -> B -> C -> X -> Y -> Z -> 1 -> 2 -> 3"`;

exports[`BinaryTree (number) render toString correctly 2`] = `"A:1 -> B:2 -> C:3 -> X:5 -> Y:5 -> Z:5 -> 1:10 -> 2:10 -> 3:10"`;
